### DISCUSSION TITLE
A Partial Refund in A Smart Contract

### DISCUSSION THREAD

**Discussion by ghost** [Agoric/agoric-sdk#9433] [Q&A] [answered]
A smart contract receives a `zcfseat` in an offer-handler function. It uses this seat to withdraw a payment using `atomicRearrange`. Later on (before calling exit on this seat), is it possible to reverse `atomicRearrange` on this seat - or to make a partial refund? An example or a pointer would be great.

**Comment by erights** [comment]
The `atomicRearrange` only changes allocations, which are only turned into payout payments on seat exit. An `atomicRearrange` must preserve offer-safety and conservation. ***While no seat has exited or entered***, any `atomicRearrange` is reversible, since the offer safety constraints are stable and conservation is reversible. Seat entry, while technically making some rearranges irreversible, does not really threaten the reversibility you have in mind. Seat exit though turns allocations into payout payments and is of course irreversible.

Attn @Chris-Hibbert 

Why? Curious if you have something specific in mind?

**Reply by ghost** [reply]
Thanks. That sounds great. Any pointers to a mechanism to arrange such partial or full reversal before seat is exited within a smart contract?

I also noticed that a user seat is payout is called after it has exited; it felt a bit unnatural for a moment.

specifics:
I am writing an auction contract that wants to hold on to the bid payments (in its own account and not with zoe), and then reimburses everyone except the winner at auction close meanwhile maybe charging a small bidding fee.

**Reply by erights** [reply]
If the auction contract holds the bids in its own accounts, what would prevent it from running off with all the bids and not reimbursing any of the bidders?

Zoe holds the bids in its own escrow purses exactly so it can provide the bidders with offer-safety, even if the auction contract misbehaves.

**Reply by ghost** [reply]
It is just a matter of choice I guess but to answer your question, auctioneer may have their repute at stake while bidders generally do not. So, it is not the totally the trust-less world that Zoe offer-safety caters to generally.

**Comment by Chris-Hibbert** [comment]
We've written several auction contracts, and it can be written either so the auction collects all the bids up front (so the bidders don't get offer safety, but the auctioneer has certainty of payment) or so the bidders get offer safety.

In the latter case, the auctioneer has to be robust to offers being withdrawn during the auction.

If the auctioneer also wants to charge a fee for bidding, it just has to reject offers that don't unconditionally pay the fee. (i.e. if their `want` says they should get their entire bid back if they don't win, then the auctioneer just returns the money without entertaining their bid.)

**Reply by ghost** [reply]
@Chris-Hibbert can you please point me to an example of the former kind. Also, is there an API for partial/full reversal of atomicTransfer/atomicRearrange? Thanks.

**Reply by Chris-Hibbert** [reply, marked as answer]
I don't see any current contracts that do this, but there's a support library in `zoe/src/contracts/auction` that shows how it would be done. [This sell method](https://github.com/Agoric/agoric-sdk/blob/8830a6eb6b178cbc3ebaae6c20ed550d5bf9eb08/packages/zoe/src/contracts/auction/index.js#L141) ensures that the seller can't exit early
```
  const sell = seat => {
    assertProposalShape(seat, {
      give: { Asset: null },
      want: { Ask: null },
      // The auction is not over until the deadline according to the
      // provided timer. The seller cannot exit beforehand.
      exit: { waived: null },
    });
    // Save the seat for when the auction closes.
    sellSeat = seat;

    // The bid invitations can only be sent out after the assets to be
    // auctioned are escrowed.
    return Far('offerResult', { makeBidInvitation, getSessionDetails });
  };
```

There's no explicit API for reversing `atomicRearrange`. (Don't use `atomicTransfer`, please.) Nothing keeps track of the particular sequence of `rearrange`s that result in the current allocation. You can look at the `offer` to find the initial allocation, and offerSafety will always allow you to restore that if the resources are available.

**Reply by ghost** [reply]
Thanks. That helps!

**Comment by erights** [comment]
> In the latter case, the auctioneer has to be robust to offers being withdrawn during the auction.

That is generally what we've done, and I do recommend it. But worth noting that the auction can insist on exit terms, such as a deadline shortly after the auction closes, that would enable it to assume bids remain during the auction.
